Scalastyle inspections generator

mwz opened the pull request:
This is still a WIP, but I'm opening it for some early feedback if you fancy reviewing it @BalmungSan. 👋

I've captured some of my questions and the things that are left to do in TODO comments, perhaps you might have some thoughts on those.

I'll also add some unit tests in a bit.

Relates to #35.

BalmungSan commented on project/ScalastyleInspectionsGenerator.scala:
```diff
@@ -0,0 +1,125 @@
+      case inspection =>
+        // TODO: Do we want to format the text with code examples?
+        // TODO: Is there a better way of embedding multi-line text?
+        val extraDescription = inspection.extraDescription.map(s => "\"\"\"" + s + "\"\"\"")
```
Maybe ` s""""$s""""` is just sufficient?

mwz replied:
Thanks for all of your suggestions - I guess another alternative would be to convert new lines into `<br>`s and that way we could just wrap the whole text in single quotes, but I'm not going to worry about it for now. As I mentioned in the comment above, I'll come back to this in a separate PR when I can see how the raw formatting looks in the UI.

BalmungSan replied:
ok

mwz commented on project/ScalastyleInspectionsGenerator.scala:
```diff
@@ -0,0 +1,125 @@
+           |  justification = $justification,
+           |  defaultLevel = ${inspection.defaultLevel}
+           |)
+         """.stripMargin
```
This code will get compiled, so we don't really care about extra whitespaces or new lines, etc.

BalmungSan replied:
Oh I see, ok.

mwz commented on project/ScalastyleInspectionsGenerator.scala:
```diff
@@ -0,0 +1,125 @@
+  private[scalastyle] def transform(source: Tree, inspections: Seq[ScalastyleInspection]): Tree = {
+    val stringified: Seq[String] = inspections.collect {
+      case inspection =>
+        // TODO: Do we want to format the text with code examples?
```
Sonar allows rules to have the description written in html and our existing implementation of the Scalastyle module re-formats the description into html markup when reading Scalastyle config. It does that so the example code snippets can be wrapped in `<p>` and `<pre>` tags, so I think that we should do something similar, but I'll leave it for now until we can actually create the rules and see how the description looks without any extra formatting. Most likely I'll need to come back to it and add some formatting.

BalmungSan commented on project/ScalastyleInspectionsGenerator.scala:
```diff
@@ -0,0 +1,125 @@
+    val generatedInspections: Seq[ScalastyleInspection] = extractInspections(inspections, docs, config)
+
+    // Load the template file from ./project/ScalastyleInspections.scala.
+    val projectDir = new File(baseDirectory.value, "project")
```
Oh, sorry...
I confused it with `java.util.Path`
However you could still make it in one line, but maybe it becomes so long 😉 

```scala
val templateFile = new File(new File(baseDirectory.value, "project"), "ScalastyleInspections.scala")
```

mwz replied:
yes it's a little bit less readable which is why I split it over two lines

BalmungSan commented on project/ScalastyleInspectionsGenerator.scala:
```diff
@@ -0,0 +1,150 @@
+      extraDescription = doc.flatMap(node => (node \ "extra-description").map(_.text.trim).headOption)
+      justification = doc.map(node => (node \ "justification").text.trim)
+      defaultLevel = Level(inspection \@ "defaultLevel")
+      params = (inspection \\ "parameter").map { param =>
```
:+1:

mwz commented:
Ok, that's all I wanted to do as a part of this PR, but I'm struggling to figure out how to unit test the code since it's in the meta-build, have you had any luck with this before @BalmungSan?

BalmungSan commented:
Sadly, not really...
As I told you before, I haven't used scala meta nor scala macros. _(and for now I'll wait for the new Scala 3 macros system to learn about meta programming, but that's another story)_.

However what you can do is write tests for the generated file, that is testing the properties of the inspections. If you recall, that's what we did for the scapegoat inspections generator.

BalmungSan commented on project/src/main/scala/ScalastyleInspections.scala:
```diff
@@ -0,0 +1,36 @@
+/*
```
Hey, it seems to me that you duplicate this file instead of renaming it from `project/ScalastyleInspections.scala` to this _(and removing the package)_.

I checked in your branch, and indeed you have this file in the [old location](https://github.com/mwz/sonar-scala/blob/scalastyle-inspections-gen/project/ScalastyleInspections.scala) too.

mwz replied:
oh sorry, I'll get rid of the old file. It's tricky because even though I use this file as a template, I also want it to be compiled along with the other one so I can be sure it is correct. I also accidentally removed the package from the file, I'll fix that too.

BalmungSan commented on project/src/test/scala/ScalastyleInspectionsGeneratorTest.scala:
```diff
@@ -0,0 +1,141 @@
+    val inspections = XML.load(fromResource("scalastyle/definition.xml").bufferedReader) \\ "checker"
+    //XML.load(this.getClass.getResourceAsStream("scalastyle/definition.xml")) \\ "checker"
+    val docs = XML.load(fromResource("scalastyle/documentation.xml").bufferedReader) \\ "check"
+    // XML.load(this.getClass.getResourceAsStream("scalastyle/documentation.xml")) \\ "check"
```
Personally I preffer to use `scala.io.Source` than `this.getClass.getResourceAsStream`.
However, I would rather just import **Source** and call it using `Source.fromResource`, also we should call bufferedReader with parentheses _()_ as stated in the [scaladoc](https://www.scala-lang.org/api/current/scala/io/BufferedSource.html#bufferedReader():java.io.BufferedReader).

mwz replied:
Ok, I'll stick to `scala.io.Source`, but I didn't import it because it collides with `scala.meta.Source` and I didn't want to rename them - I've imported `scala.io.Source.fromResource` instead, are you ok with that?

> we should call bufferedReader with parentheses () as stated in the scaladoc.

👍

BalmungSan commented on project/src/test/scala/ScalastyleInspectionsGeneratorTest.scala:
```diff
@@ -0,0 +1,141 @@
+
+    val result = ScalastyleInspectionsGenerator.transform(source, inspections)
+
+    result.structure == expected.parse[Source].get.structure shouldBe true
```
I may be wrong, but it is not this redundant?
I mean `shoudlBe` should just call `==` if not especial Equality is present, right?

mwz replied:
you're right - it wasn't necessary, I've replaced it with just `shouldBe`

BalmungSan commented on project/src/test/resources/scalastyle/documentation.xml:
```diff
@@ -0,0 +1,61 @@
+            ]]>
+        </example-configuration>
+    </check>
+</scalastyle-documentation>
```
Missing EOL here 😉

BalmungSan commented on project/src/test/scala/ScalastyleInspectionsGeneratorTest.scala:
```diff
@@ -0,0 +1,141 @@
+        "Checks that the ScalaDoc on documentable members is well-formed",
+        Some(
+          """Ignore tokens is a comma separated string that may include the following : PatDefOrDcl (variables), TmplDef (classes, traits), TypeDefOrDcl (type definitions), FunDefOrDcl (functions)
+            |            Supported indentation styles are "scaladoc" (for ScalaDoc-style comments, with two spaces before the asterisk), "javadoc" (for JavaDoc-style comments, with a single space before the asterisk) or "anydoc" to support any style (any number of spaces before the asterisk). For backwards compatibility, if left empty, "anydoc" will be assumed.""".stripMargin
```
Uhm... We should find a way to fix this bad identatio, maybe replacing all whitespaces after a new line with the empty char while extracting or transforming the inspections will be sufficient?

Or are these extra espaces also removed when compiling the generated code?

mwz replied:
Yes, I agree - as I said before, this will probably need some reformatting similarily to what the existing module does, but I'll come back to this separately once I can create those rules in Sonar to see what the formatting looks like in the UI.

BalmungSan replied:
Ok.

BalmungSan commented:
Hi @mwz, I liked the unit tests in general terms, just a few things that I leave in my comments.

However, I have two questions...

1. Why removing the package?, not mad at all, just curious.

2. What do you think about testing the generated file too?, I mean maybe its redundant of what you already done, but with the difference that it will be done with the real scalastyle rules.

mwz commented on project/ScalastyleInspections.scala:
```diff
@@ -0,0 +1,36 @@
+
+import org.scalastyle._
+
+case class ScalastyleInspection(
```
I generally agree with you, but it turns out there are way too many imports from Scalastyle so IntelliJ just automatically optimizes it into a wildcard import.

BalmungSan replied:
Ok, Is this code analyzed in sonar? _(I think not)_
If so, we should add the ignore comment.

mwz replied:
No, this sits in the project folder, so this doesn't get analysed by Sonar.

mwz commented:
> Why removing the package?, not mad at all, just curious.

I removed the package from the `ScalastyleInspectionsGenerator` because it's just a build code which doesn't need a proper namespace, but I've also accidentally removed the package from the `ScalastyleInspections` file, which does need the namespace as we'll use it from the main code base - I'll fix that.

> What do you think about testing the generated file too?, I mean maybe its redundant of what you already done, but with the difference that it will be done with the real scalastyle rules.

There is never enough unit tests, so I'm definitely up for doing those as a part of the rules repo.

Thanks for your review, @BalmungSan - I've addressed your comments in the latest commits.

mwz commented:
I've also added missing parameter labels and descriptions, I think it should be all this time.

BalmungSan commented on project/src/main/scala/ScalastyleInspections.scala:
```diff
@@ -16,6 +16,9 @@
+
 import org.scalastyle._
 
 case class ScalastyleInspection(
```
I'm sorry I did not notice it before...
But, we should make those two case classes `final`, and also we should make them only visible in the scalastyle package.

However, as you said before this needs refactor and you want to see how this behave first. So, we could omit this for now.

BalmungSan replied:
> "funny how rarely you see people make their case classes final."

I think Scala 3, will make them final by default.

> "The problem is that probably the recommended quality profile code is going to live on the same level as scalastyle and scapegoat, so if we want to have access to those inspections there, we won't be able to make them private to their own packages..."

Ah, I see, well nothing to do then.
